### PR TITLE
cloudstack: Setup temporary SSH keypair

### PR DESCRIPTION
**Pull request by rickard-von-essen**
Also adds printout of the insance password in debug mode.

This contains a _backwards incompatible_ change in the uncommon case that the source image allowed SSH auth with password but not with keypair. This should be noted in the changelog.

### CODE REVIEW

File: builder/cloudstack/builder.go [hunk 1]
```diff
@@ -61,8 +63,17 @@ func (b *Builder) Run(ui packer.Ui, hook packer.Hook, cache packer.Cache) (packe
 			HTTPPortMin: b.config.HTTPPortMin,
 			HTTPPortMax: b.config.HTTPPortMax,
 		},
+		&stepKeypair{
+			Debug:            b.config.PackerDebug,
+			DebugKeyPath:     fmt.Sprintf("cs_%s.pem", b.config.PackerBuildName),
+			SSHAgentAuth:     b.config.Comm.SSHAgentAuth,
+			TemporaryKeyPair: b.config.TemporaryKeypair,
+			KeyPair:          b.config.Keypair,
+			PrivateKeyFile:   b.config.Comm.SSHPrivateKey,
```
Comment (svanharmelen) [inline builder/cloudstack/builder.go]: In the other steps these entries are ordered alphabetically...

File: builder/cloudstack/builder.go [hunk 2]
```diff
@@ -61,8 +63,17 @@ func (b *Builder) Run(ui packer.Ui, hook packer.Hook, cache packer.Cache) (packe
...
+		},
 		&stepCreateInstance{
-			Ctx: b.config.ctx,
+			Debug: b.config.PackerDebug,
+			Ctx:   b.config.ctx,
```
Comment (svanharmelen) [inline builder/cloudstack/builder.go]: Same for these entries (alphabetical ordering).

File: builder/cloudstack/config.go [hunk 3]
```diff
@@ -34,6 +34,7 @@ type Config struct {
 	Hypervisor        string   `mapstructure:"hypervisor"`
 	InstanceName      string   `mapstructure:"instance_name"`
 	Keypair           string   `mapstructure:"keypair"`
+	TemporaryKeypair  string   `mapstructure:"temporary_keypair"`
```
Comment (svanharmelen) [inline builder/cloudstack/config.go]: If I understand it correctly this field is not to supply a temporary keypair, but the name the Packer should use when creating a temporary keypair. I think this naming is a bit misleading.

File: builder/cloudstack/config.go [hunk 4]
```diff
@@ -120,6 +121,15 @@ func NewConfig(raws ...interface{}) (*Config, error) {
 		c.TemplateDisplayText = c.TemplateName
 	}
 
+	// If we are not given an explicit keypair or ssh_private_key_file, then create
+	// a temporary one, but only if the temporary_keypair has not been provided and
+	// we are not using ssh_password.
```
Comment (svanharmelen) [inline builder/cloudstack/config.go]: Hmm... Should this comment be:

```
// If we are not given an explicit keypair or ssh_private_key_file, then create
// a temporary keypair name, but only if the temporary_keypair_name has not
// been provided and we are not using ssh_password.
```

Of do I fail to understand what you are trying to do here?

File: builder/cloudstack/step_keypair.go [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+
+	if s.PrivateKeyFile != "" {
+		privateKeyBytes, err := ioutil.ReadFile(s.PrivateKeyFile)
+		if err != nil {
+			state.Put("error", fmt.Errorf(
+				"Error loading configured private key file: %s", err))
+			return multistep.ActionHalt
+		}
+
+		state.Put("keypair", s.KeyPair)
```
Comment (svanharmelen) [inline builder/cloudstack/step_keypair.go]: Shouldn't we `ui.Say` something above this line as well, about what we'll be using?

File: builder/cloudstack/step_keypair.go [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+	}
+
+	if s.SSHAgentAuth && s.KeyPair != "" {
+		ui.Say(fmt.Sprintf("Using SSH Agent for existing keypair %s", s.KeyPair))
+		state.Put("keypair", s.KeyPair)
+		return multistep.ActionContinue
+	}
+
+	if s.TemporaryKeyPair == "" {
+		ui.Say("Not using temporary keypair")
```
Comment (svanharmelen) [inline builder/cloudstack/step_keypair.go]: I would rephrase this to something like: `Not using a keypair` It feels a bit wierd that is you did not specify a temporary keypair, we explicitly say we are not going to use it as we wouldn't expect anything else right?

File: builder/cloudstack/step_keypair.go [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+		state.Put("keypair", "")
+		return multistep.ActionContinue
+	}
+
+	client := state.Get("client").(*cloudstack.CloudStackClient)
+
+	ui.Say(fmt.Sprintf("Creating temporary keypair: %s ...", s.TemporaryKeyPair))
+	p := client.SSH.NewCreateSSHKeyPairParams(
+		s.TemporaryKeyPair,
+	)
```
Comment (svanharmelen) [inline builder/cloudstack/step_keypair.go]: This can be one line, it's still short enough :wink:

File: builder/cloudstack/step_keypair.go [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+		err := fmt.Errorf("The temporary keypair returned was blank")
+		state.Put("error", err)
+		ui.Error(err.Error())
+		return multistep.ActionHalt
+	}
+
+	ui.Say(fmt.Sprintf("Created temporary keypair: %s", s.TemporaryKeyPair))
+
+	// If we're in debug mode, output the private key to the working
+	// directory.
```
Comment (svanharmelen) [inline builder/cloudstack/step_keypair.go]: This can also be just one line, still way less then 80 chars right?

File: builder/cloudstack/step_keypair.go [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+	state.Put("keypair", s.TemporaryKeyPair)
+	state.Put("privateKey", keypair.Privatekey)
+
+	return multistep.ActionContinue
+}
+
+func (s *stepKeypair) Cleanup(state multistep.StateBag) {
+	if !s.doCleanup {
+		return
+	}
```
Comment (svanharmelen) [inline builder/cloudstack/step_keypair.go]: Why not just check if `s. TemporaryKeyPair` has a value? Seems a bit cleaner.

File: builder/cloudstack/config.go (hunk 3 above)
Comment (svanharmelen) [inline builder/cloudstack/config.go]: Ah, the docs say it correctly... But that means this is wrong and this should be `TempKeypairName` and `temporary_keypair_name`.

File: builder/cloudstack/config.go [hunk 10]
```diff
@@ -120,6 +121,15 @@ func NewConfig(raws ...interface{}) (*Config, error) {
...
+	if c.Keypair == "" && c.TemporaryKeypair == "" &&
+		c.Comm.SSHPrivateKey == "" && c.Comm.SSHPassword == "" {
+
```
Comment (svanharmelen) [inline builder/cloudstack/config.go]: I would remove this blank line...

File: builder/cloudstack/step_keypair.go [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+		// Chmod it so that it is SSH ready
+		if runtime.GOOS != "windows" {
+			if err := f.Chmod(0600); err != nil {
+				err := fmt.Errorf("Error setting permissions of debug key: %s", err)
+				state.Put("error", err)
+				ui.Error(err.Error())
+				return multistep.ActionHalt
+			}
+		}
+	}
```
Comment (svanharmelen) [inline builder/cloudstack/step_keypair.go]: Does it make sense to fail the whole packer build when a debug action failed?

File: builder/cloudstack/step_keypair.go [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+
+	// we created a temporary key, so remember to clean it up
+	s.doCleanup = true
```
Comment (svanharmelen) [inline builder/cloudstack/step_keypair.go]: Seems a bit of an ugly why to do this... IMO that is.

**Comment by svanharmelen** [comment]
I just noticed that in #5175 you used a value `create_security_group bool` to determine if you wanted to use a temp security group. Could that be a better approach for this one as well?

I don't really see the added value of being able to be able to supply a name for a temp keypair. Just being able to tell Packer that you want to use a temp keypair seems to make a bit more sense to me.

File: builder/cloudstack/step_keypair.go (hunk 12 above)
Comment (rickard-von-essen) [inline builder/cloudstack/step_keypair.go]: This is a bit of legacy from the AWS builder. The logic is a bit different there so maybe it should be cleaned up.. I'll have a look.

**Comment by rickard-von-essen** [comment]
> I just noticed that in #5175 you used a value create_security_group bool to determine if you wanted to use a temp security group. Could that be a better approach for this one as well?

SG's are optional but a temporary keypair should be the default, unless you use a specified allready existing keypair, or password/ssh key that is allready "built" into the source image etc.

> I don't really see the added value of being able to be able to supply a name for a temp keypair. Just being able to tell Packer that you want to use a temp keypair seems to make a bit more sense to me.

It's probably a bit overkill but it follows the AWS builders and might be usefull if you want to name all resources per team etc.

File: builder/cloudstack/config.go (hunk 3 above)
Comment (rickard-von-essen) [inline builder/cloudstack/config.go]: Yes I was hesitating about naming it `temporary_keypair` or `temporary_keypair_name`. I'll change, the later is better.

File: builder/cloudstack/config.go (hunk 4 above)
Comment (rickard-von-essen) [inline builder/cloudstack/config.go]: Yeah that didn't make any seense.

File: builder/cloudstack/step_keypair.go (hunk 5 above)
Comment (rickard-von-essen) [inline builder/cloudstack/step_keypair.go]: This step is in practice skipped if using a provided keypair.

File: builder/cloudstack/step_keypair.go (hunk 6 above)
Comment (rickard-von-essen) [inline builder/cloudstack/step_keypair.go]: 👍

File: builder/cloudstack/step_keypair.go (hunk 11 above)
Comment (rickard-von-essen) [inline builder/cloudstack/step_keypair.go]: Defacto standard. I can't say one is better than the other so we arbitrary use one.

**Comment by rickard-von-essen** [comment]
Ready for re-review. Thanks for the quick feedback.

**Comment by svanharmelen** [comment]
LGMT, but maybe you can kick the tests? I noticed I can't restart the job in AppVeyor...
